core/src/memory.rs

```diff
@@ -144,6 +144,11 @@ impl Memory {
 		len: U256,
 		data: &[u8],
 	) -> Result<(), ExitFatal> {
+		if len.is_zero() {
```

Comment:
This is actually not necessary for semantics, since below we call `set`, and within that function it checked the len is zero case.

I understand the concern of a short-cut here since we do not need to do additional checks below, but there are indeed several things:

* If we add this check, it should be below the `ErrorFatal::NotSupported` check but not above. The `ErrorFatal` type is a unique error type not covered in spec but useful in practice for non-metered cases -- it ensures we can always handle the addressing spaces since EVM supports 256-bit memory, but in reality all current EVM executors can only support up to 64-bit memory running on x86_64.
* The len zero condition should be triggered extremely rarely since it's basically a no-op. And a well-optimized EVM contract should not have triggered it.
* Our memory code `set` and `copy_large` is already quiet complex and it happened that some trivial changes in the past caused compliance issue. We should really not add additional "patches" unless necessary and instead prefer to fully refactor those two functions.

With all the above said, my current recommendation is to remove this change. We then spend some time refactor `set` and `copy_large` to make it less error-prone and then consider to add those sugar-sized optimizations.

Reply:
I added this change to pass this [eth json test](https://github.com/ethereum/tests/commit/17f7e7a6c64bb878c1b6af9dc8371b46c133e46d) (you can see it linked from [the commit where the change was introduced](https://github.com/rust-blockchain/evm/pull/40/commits/3e081dc934d12d4405c082bcf2ff968ec8cc382d)).

That eth json test was added before Berlin tests so when we updated the evm tests repo to point at a later version of ethereum tests this change was needed to have them all pass.

I agree it's a little hacky, but I do not think we should block merging Berlin changes to refactor the memory functions. We could create an issue to do that refactoring and come back to it after Berlin changes are merged.

Reply:
We don't need that now though -- till the moment we update jsontests.

Alternatively, you can add comments to this line explaining why this is introduced now, and that this is a temporary solution and will eventually be refactored. The priority here should be to avoid more confusions.

Reply:
Added a comment in https://github.com/rust-blockchain/evm/pull/40/commits/9bee750378387dd8eaa7bb774547fa89f16d68be

> till the moment we update jsontests.

Are we not going to update the jsontests along side this PR?

Reply:
That's what I am confused about, as jsontests will be updated immediately after.